docs/proposals/13-automated-seed-management.md

```diff
@@ -0,0 +1,289 @@
+* From the remaining seeds, determine candidates that are considered best based on their region, by using a strategy that can be either "same region" or "minimal distance".
+* Among these candidates, choose the one with the least number of shoots.
+
+This scheduling algorithm should be adapted in order to properly take into account resources capacity and requests. As a first step, during the filtering stage, any seeds that would exceed their capacity for shoots, or their capacity for any resources requested by the shoot, should simply be filtered out and not considered during the next stages.
+
+Later, the scheduling algorithm could be further enhanced by replacing the step in which the region strategy is applied by a scoring step similar to the one in [Kubernetes Scheduler](https://kubernetes.io/docs/concepts/scheduling-eviction/kube-scheduler/). In this scoring step, the scheduler would rank the remaining seeds to choose the most suitable shoot placement. It would assign a score to each seed that survived filtering based on a list of scoring rules. These rules might include for example `MinimalDistance` and `SeedResourcesLeastAllocated`, among others. Each rule would produce its own score for the seed, and the overall seed score would be calculated as a weighted sum of all such scores. Finally, the scheduler would assign the shoot to the seed with the highest ranking. 
+
+## Creating, Deleting, and Updating Seeds Declaratively
+
+When all or most of the existing seeds are near capacity, new seeds should be created in order to accommodate more shoots. Conversely, sometimes there could be too many seeds for the number of shoots, and so some of the seeds could be deleted to save resources. Currently, the process of creating a new seed involves a number of manual steps, such as creating a new shoot that meets certain criteria, and then registering it as a seed in Gardener. This could be automated to some extent by [annotating a shoot with the `use-as-seed` annotation](../usage/shooted_seed.md), in order to create a "shooted seed". However, adding more than one similar seeds still requires manually creating all needed shoots, annotating them appropriately, and making sure that they are successfully reconciled and registered.
+
+To create, delete, and update seeds effectively in a declarative way and allow auto-scaling, a "creatable seed" resource along with a "set" and a "deployment"  of such creatable seeds should be introduced, similar to Kubernetes `Pod`, `ReplicaSet`, and `Deployment` (or to MCM `Machine`, `MachineSet`, and `MachineDeployment`) resources. With such resources (and their respective controllers), creating a new seed based on a template would become as simple as increasing the `replicas` field in the "set" resource, and updates to seeds could be performed in a rolling manner via the "deployment" resource. 
```

Comment:
I don't fully understand the point regarding the replicas. If the replicas field is increased, then a new shooted seed, identical with the already existing ones, is created and registered. This is also explained in the respective section. Do you see any issues with this? 

If the replicas field is decreased, then a seed that has no shoots is deleted. If there is no such seed, initially this might end with an error. As an alternative, we may initially forbid decreasing the replicas. In the future, once we are more confident with migration / draining seeds, decreasing replicas could be feasible even if this would mean migrating shoots away from a seed to be deleted. If you agree, I should perhaps add this explanation to the document to make it more clear.

> What about having ShootedSeedClass or ShootedSeedTemplate resource, and then concrete ShootedSeed resources referencing such classes/templates?

What would be the advantage of having them, compared with just having the template in the ShootedSeed as mentioned in the respective section?

> What is meant by this, do you have one or two example use-cases?

No, I don't have a concrete use case. I added the entire `ShootedSeedDeployment` because there was consensus in the team meeting that this might actually be needed and valuable. It could also be removed entirely (or perhaps referred to as the subject of a future GEP), as it's not critical for autoscaling. @mvladev, could you help here, I think you had concrete points in mind during the meeting?

Reply:
> I don't fully understand the point regarding the replicas. If the replicas field is increased, then a new shooted seed, identical with the already existing ones, is created and registered. This is also explained in the respective section. Do you see any issues with this?

Scaling-out is fine, sure, but scaling-in is problematic as you write in the next paragraph. In most cases there won't be a seed with no shoots, so scaling-in would indeed require to drain the least utilized seed so that it can be deleted. This drain/control plane migration is something for which we don't any experience yet. If we forbid decreasing the replicas then we cannot even (manually) get rid of a seed anymore, right?

I think the concerns arise out of the fact that shooteds/seeds are heavy-weight and cannot be compared to deployments/replicasets too much due to this. There is no rolling update, there is no rollback, many of the concepts for normal deployments/replicasets cannot be applied because they rely on the fact that you can easily terminate a pod an restart it somewhere else.

Hence, I made the proposal with the class/template resource. I don't know if they are a good solution, but at least they don't pretend to have features comparable to those from Kubernetes deployments/replicasets.

Reply:
To get rid of a seed, you could still manually migrate or delete its shoots, and then decrease the replicas field. I don't think this is any worse than what we have today, correct?

I fully agree with the concerns that seeds are heavyweight and not directly comparable to pods. I still need a bit more help to understand how introducing `ShootedSeedClass` fits better - could you please elaborate? I need this also in order to better motivate such a design when I introduce it in the GEP.

More concretely, I would propose to drop `ShootedSeedDeployment` entirely and work only with "sets" if there are no strong reasons to keep it.

Reply:
After a discussion with @rfranzke we agreed to keep the `ShootedSeedSet` more or less as it was originally proposed, and to add some details regarding how the challenges involved in deleting seeds could be addressed. We also agreed to remove the `ShootedSeedDeployment`. The document has been revised accordingly.